pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,393 @@
+	if pod.DeletionTimestamp == nil {
+		for _, podClaim := range pod.Spec.ResourceClaims {
+			if podClaim.Source.Template != nil {
+				// It has at least one inline template, work on it.
+				ec.queue.Add(podKeyPrefix + pod.Namespace + "/" + pod.Name)
+				break
+			}
+		}
+	}
+}
+
+func (ec *resourceClaimController) onResourceClaimDelete(obj interface{}) {
+	claim, ok := obj.(*v1.ResourceClaim)
+	if !ok {
+		return
+	}
+
+	// Someone deleted a ResourceClaim, either intentionally or
+	// accidentally. If there is a pod referencing it because of
+	// an inline resource, then we should re-create the ResourceClaim.
+	// The common indexer does some prefiltering for us by
+	// limiting the list to those pods which reference
+	// the ResourceClaim.
+	objs, err := ec.podIndexer.ByIndex(podResourceClaimIndex, fmt.Sprintf("%s/%s", claim.Namespace, claim.Name))
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("listing pods from cache: %v", err))
+		return
+	}
+	for _, obj := range objs {
+		ec.enqueuePod(obj)
+	}
+}
+
+func (ec *resourceClaimController) Run(ctx context.Context, workers int) {
+	defer runtime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	klog.Infof("Starting ephemeral volume controller")
+	defer klog.Infof("Shutting down ephemeral volume controller")
+
+	if !cache.WaitForNamedCacheSync("ephemeral", ctx.Done(), ec.podSynced, ec.claimsSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, ec.runWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (ec *resourceClaimController) runWorker(ctx context.Context) {
+	for ec.processNextWorkItem(ctx) {
+	}
+}
+
+func (ec *resourceClaimController) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := ec.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	logger := klog.LoggerWithValues(klog.FromContext(ctx), "key", key)
+	ctx = klog.NewContext(ctx, logger)
+	err := ec.syncHandler(ctx, key.(string))
+	if err == nil {
+		ec.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	ec.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler is invoked for each work item which might need to be processed.
+// If an error is returned from this function, the item will be requeued.
+func (ec *resourceClaimController) syncHandler(ctx context.Context, key string) error {
+	sep := strings.Index(key, ":")
+	if sep < 0 {
+		return fmt.Errorf("unexpected key: %s", key)
+	}
+	prefix, object := key[0:sep+1], key[sep+1:]
+	namespace, name, err := cache.SplitMetaNamespaceKey(object)
+	if err != nil {
+		return err
+	}
+
+	switch prefix {
+	case podKeyPrefix:
+		return ec.syncPod(ctx, namespace, name)
+	case claimKeyPrefix:
+		return ec.syncClaim(ctx, namespace, name)
+	default:
+		return fmt.Errorf("unexpected key prefix: %s", prefix)
+	}
+
+}
+
+func (ec *resourceClaimController) syncPod(ctx context.Context, namespace, name string) error {
+	logger := klog.FromContext(ctx)
```

Comment:
This is a bit redundant because earlier we did `klog.LoggerWithValues(klog.FromContext(ctx), "key", key)` where the key identifies the pod.

As a developer, I wanted to see the key, but as a user it probably makes more sense to log either pod or claim. Makes some other log calls simpler, too. I switched to that.